### PR TITLE
Add request review functionality to card

### PR DESCRIPTION
**Pull request by kingraphaii**
Related issues: [please specify]

## Description:

This PR adds the "request review" button along with a minimal "alert" api for cards.

## Screenshots/videos

image

## I solemnly swear that:

- [x] My code follows the style guidelines of this project
- [x] I have merged the develop branch into my branch and fixed any merge conflicts
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested new or existing tests and made sure that they pass

### CODE REVIEW

File: backend/curriculum_tracking/models.py [hunk 1]
```diff
@@ -1667,6 +1667,34 @@ def request_user_is_assignee(self):
...
+                    user.has_perm(Team.PERMISSION_MANAGE_CARDS, team)
+                    for team in self.get_teams()
+                )
+            )
+
+            if is_assignee or is_superuser or has_manage_cards_permission:
+                return True
+
+        return False
+
```
Comment (Ngoakor12) [inline backend/curriculum_tracking/models.py]: - Should we not return early if the user is the assignee before checking anything else?
- I'm not sure this button is supposed to show for superusers or team managers. Maybe something to confirm with Sheena? I just haven't seen it in the UI so far.
- Can we maybe add a test for this method similar to the one for `request_user_can_start` would be useful too?

File: backend/frontend/views.py [hunk 2]
```diff
@@ -274,6 +277,59 @@ def action_start_card(request, card_id):
     )
 
 
+def user_can_request_review(logged_in_user):
+    if len(helpers.agile_card_reviews_outstanding(logged_in_user)):
+        return False
+    if len(helpers.pull_request_reviews_outstanding(logged_in_user)):
+        return False
```
Comment (Ngoakor12) [inline backend/frontend/views.py]: I think we can ignore these checks if we want to remind the learners to review outstanding reviews before starting in the UI like we currently are. It's probably possible but I think the button should be there for us to have an opportunity to remind them.

I don't know if I'm explaining this well or not, so let me know if it doesn't make sense

**Review by Ngoakor12 — CHANGES_REQUESTED** [review]
The `Request Review` button is showing up on all the assignee's cards in "in progress" and "review feedback". There should be an exception: topic cards that don't require a review. Those cards will have a "done" button instead.

File: backend/frontend/views.py (hunk 2 above)
Comment (kingraphaii) [reply backend/frontend/views.py]: I did not quite get you, but this isn't enforced at the model level, and as per the current setup, one shouldn't be able to move it to IR, hence its presence here. 

Also, the fact that this is on the "action" view, not the main one in charge of displaying the cards, means there's still that visual cue for the outstanding IR cards.

Let me know if we're talking about the same thing 😅

Comment (Ngoakor12) [reply backend/frontend/views.py]: Lol, let me try again.

The color being a visual cue is all well and good. I just think it is better to be explicit about why your card is not able to move to review with something like a toast or a modal. Especially for new users who just know they can move cards from 1 column to the other.

Also, my thinking was around how easy it would be to implement that "explicit UI feedback" mentioned above. My thinking is we would still have to do some checks with the above helper functions again since the ones used in the above function only checks a lot of stuff including the other permissions.

Comment (Ngoakor12) [reply backend/frontend/views.py]: I just saw the logic for the permission classes in the api_views disregard the above review, I guess? :sweat_smile:

Comment (kingraphaii) [reply backend/frontend/views.py]: Got it

File: backend/curriculum_tracking/models.py (hunk 1 above)
Comment (kingraphaii) [reply backend/curriculum_tracking/models.py]: > * I'm not sure this button is supposed to show for superusers or team managers. Maybe something to confirm with Sheena? I just haven't seen it in the UI so far.

I did this following the existing implementation here (in the current DRF api views):

```python
permission_classes=[
    (
        curriculum_permissions.IsCardAssignee
        & curriculum_permissions.AssigneeHasFinishedReviews
    )
    | HasObjectPermission(
        permissions=Team.PERMISSION_MANAGE_CARDS,
        get_objects=_get_teams_from_card,
    )
]
```

Refer to the `or` expression after the `|`

Comment (Ngoakor12) [reply backend/curriculum_tracking/models.py]: Apologies, I thought your response was related to my last comment, but now I see that it did not send. I did actually end up seeing the logic for API view access.

image

File: backend/frontend/tests/test_card_request_review_button.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        self.page.click("text=Request Review")
+
+        self.page.wait_for_load_state("networkidle")
+
+        rf_column = self.page.text_content("div#column_RF")
+        review_column = self.page.text_content("div#column_IR")
+        card_title = self.card.content_item.title
+
+        self.assertIn(card_title, review_column)
+        self.assertNotIn(card_title, rf_column)
```
Comment (Ngoakor12) [inline backend/frontend/tests/test_card_request_review_button.py]: Can you add a test for the card movements when there are outstanding reviews. I can't seem to move a card even after adding a review on an outstanding card

File: backend/frontend/views.py [hunk 4]
```diff
@@ -1,4 +1,3 @@
-import json
```
Comment (Ngoakor12) [inline backend/frontend/views.py]: This module should not be removed,  it's used by the dashboard

**Review by Ngoakor12 — CHANGES_REQUESTED** [review]

File: backend/frontend/static/frontend/utils.js [hunk 5]
```diff
@@ -5,3 +5,15 @@ function countLoadedPageItems(parentId) {
...
+
+/*
+ * Swaps in the body of error pages returned from htmx requests 
+ * for example when an action is forbidden.
+ */
+document.addEventListener("htmx:beforeOnLoad", function (event) {
+  const xhr = event.detail.xhr
+  if (xhr.status == 500 || xhr.status == 403 || xhr.status == 404) {
+    event.stopPropagation() // Tell htmx not to process these requests
+    document.children[0].innerHTML = xhr.response // Swap in body of response instead
```
Comment (sheenarbw) [inline backend/frontend/static/frontend/utils.js]: Let's say we have a page that uses htmx to fetch 10 graphs from the backend as it loads. If graph number 8 comes up with an error then the entire page will show that wrror. But the other 9 graphs could be useful.

Generally with frontend dev you want to make sure that errors are well contained. If one little widget is broken you dont want the whole page to break.   Here you seem to be explicitly breaking the whole page. 

What are you trying to achieve with this?

File: backend/frontend/templates/frontend/user/board/partial_card.html [hunk 6]
```diff
@@ -166,5 +166,17 @@
             >Start
         </button>
     {% endif %}
+    
+    {% if card.content_type_nice == "project" or card.content_type_nice == "topic" and card.topic_needs_review %}
```
Comment (sheenarbw) [inline backend/frontend/templates/frontend/user/board/partial_card.html]: We need to get rid of topic_needs_review. It's debt. Please don't build it into new logic

File: backend/frontend/views.py [hunk 7]
```diff
@@ -295,6 +297,77 @@ def action_start_card(request, card_id):
...
+            {
+                "card": card,
+                "message": "Please make sure you have reviewed assigned pull requests.",
+                "alert_type": "error",
+            },
+        )
+
+    if card.recruit_project:
+        card.recruit_project.request_review(force_timestamp=timezone.now())
+    elif card.topic_progress:
```
Comment (sheenarbw) [inline backend/frontend/views.py]: The request review button should not show up for topics. If there is a topic_progress instance it would be an error.

For topics you should just be able to say that they are finished, it's a different button

File: backend/frontend/views.py [hunk 8]
```diff
@@ -295,6 +297,77 @@ def action_start_card(request, card_id):
...
+        card.finish_topic()
+    else:
+        raise NotImplementedError()
+
+    log_creators.log_card_review_requested(card=card, actor_user=request.user)
```
Comment (sheenarbw) [inline backend/frontend/views.py:352]: it would be a bug to log that a review was requested if a topic was finished.  We need to be careful to be explicit with these logs

**Review by sheenarbw — CHANGES_REQUESTED** [review]
I just had a quick look. I didn't review your tests.

File: backend/frontend/static/frontend/utils.js (hunk 5 above)
Comment (kingraphaii) [reply backend/frontend/static/frontend/utils.js]: Had not thought about this, thank you for pointing it out. I wanted a way to communicate errors for example we have this guard on a view that gets called via ajax:

```python
@user_passes_test_or_forbidden(user_can_request_review)
def action_request_review(request, card_id):
    ...
```

if the request user fails the test, then the user won't know what happened, the error will only be visible in the browser console. 

Is this something we should worry about on this PR or we can leave it out for later?

Comment (sheenarbw) [reply backend/frontend/static/frontend/utils.js]: That's a good point. Bt I do think having it fail in this way is much better than breaking hydration. We use hydration already and I'll sure we'll use it more.

I think the thing to do is treat things differently per error. Eg a 500 error is something that really shouldn't happen, it means our backend code fell over. If those happen we can just have an alert saying "something went wrong" or something like that. 

500 errors will get picked up in our error logs and I'll get an email if there are any new ones so things will get fixed.

404s really shouldn't happen in an xhr request because we'll only really be fetching things we are allowed to fetch, with known ids. This would also actually be a bug I think. Same for 403s.  

A 403/4 can also come from a user navigating to a page they are not allowed to see but that should not be handled here.

I think for now it's ok to remove this event handler.

**Review by Ngoakor12 — APPROVED** [review]
Changes addressed and the tests are passing

File: backend/frontend/static/frontend/user_board.js [hunk 9]
```diff
@@ -21,3 +21,51 @@ function moveCardToCorrectColumn(cardId, column) {
...
   }
 }
+
+
+/*
+This inserts an alert above the card being acted on.
+It's used to show any errors or success messages that 
+come back from the server.
+*/
+function showCardAlert({ cardId, message, alertType = "info", persist = true }) {
```
Comment (sheenarbw) [inline backend/frontend/static/frontend/user_board.js]: This is similar to some code from Ngoako that I just reviewed.  You shouldn't be implementing the same thing in 2 different PRs.

File: backend/frontend/templates/frontend/user/board/js_exec_action_card_moved.html [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{% include "frontend/user/board/partial_card.html" %}
+
+<script>
```
Comment (sheenarbw) [inline backend/frontend/templates/frontend/user/board/js_exec_action_card_moved.html:3]: Hmm.. same comment as above. Also there is already a file that does this, now the code isn't DRY.

File: backend/frontend/templates/frontend/user/board/js_exec_action_show_card_alert.html [hunk 11]
```diff
@@ -0,0 +1,5 @@
+{% include "frontend/user/board/partial_card.html" %}
+
+<script>
+  showCardAlert({cardId: "{{card.id}}", message: "{{ message }}", alertType: "{{ alert_type }}", persist: {{ persist|default:"false" }} });
```
Comment (sheenarbw) [inline backend/frontend/templates/frontend/user/board/js_exec_action_show_card_alert.html]: feedback given to Ngoako already

File: backend/frontend/views.py [hunk 12]
```diff
@@ -295,6 +298,75 @@ def action_start_card(request, card_id):
     )
 
 
+def user_can_request_review(logged_in_user):
```
Comment (sheenarbw) [inline backend/frontend/views.py]: I think we need a naming convention for these permission functions. It's really not clear what the intention of this code is until I go search for where you used it. 

Can you think of a better naming convention? It's also not clear from the name that a card_id is required here. 

Ideally I should know what a function is supposed to do just by looking at the name.

File: backend/frontend/views.py [hunk 13]
```diff
@@ -295,6 +298,75 @@ def action_start_card(request, card_id):
...
+    """The card is in progress or review feedback and the user has chosen to request review"""
+    card = get_object_or_404(AgileCard, id=card_id)
+
+    if len(helpers.agile_card_reviews_outstanding(request.user)):
+        return render(
+            request,
+            "frontend/user/board/js_exec_action_show_card_alert.html",
+            {
+                "card": card,
+                "message": "You have outstanding card reviews.",
```
Comment (sheenarbw) [inline backend/frontend/views.py]: This kind of logic also exists in Ngoako's PR. It wont be DRY. We need functionality to be centralised and referenced, not repeated.

File: backend/frontend/templates/frontend/user/board/page.html [hunk 14]
```diff
@@ -1,14 +1,37 @@
 {% extends "frontend/user/base.html" %}
 
+{% block head %}
+    {{ block.super }}
+    <style>
+        @keyframes shake {
+            10%, 90% {
```
Comment (sheenarbw) [inline backend/frontend/templates/frontend/user/board/page.html]: also the same across PRs

File: backend/frontend/views.py [hunk 15]
```diff
@@ -295,6 +298,75 @@ def action_start_card(request, card_id):
...
 
 
+def user_can_request_review(logged_in_user):
+    request = get_current_request()
+    card_id = request.resolver_match.kwargs.get("card_id")
+
+    card = get_object_or_404(AgileCard, pk=card_id)
+    card_assignees = card.assignees.all()
+
+    if logged_in_user in card_assignees:
```
Comment (sheenarbw) [inline backend/frontend/views.py]: This functionality seems to be repeated in AgileCard.request_user_can_request_review

**Review by sheenarbw — CHANGES_REQUESTED** [review]
It might make sense for you or Ngoako to pause while we get one thing merged. There is a lot of stuff that is repeated across PRs and it's going to get messy to merge these things

**Comment by kingraphaii** [comment]
@Ngoakor12 

As suggested in previous reviews, I'm proposing we use the check_[verb]_[optional_qualifier] for naming our permission checking functions.

Example:
```python
def check_user_can_request_review_on_card(logged_in_user):
    return [bool]

@user_passes_test_or_forbidden(check_user_can_request_review_on_card)
def decorated_view():
    ...
```
I've used it on the "request_review" action only until we all agree on a naming style for them.

Sheena also mentioned how we do not make it known that some of these expect a "card_id" to be present in the `kwargs,` but I haven't been able to come up with a clean solution so far; feel free to suggest a better way to handle this.

File: backend/frontend/tests/test_card_request_review_button.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        self.page.click("text=Request Review")
+
+        self.page.wait_for_load_state("networkidle")
+
+        ip_column = self.page.text_content("div#column_IP")
+        review_column = self.page.text_content("div#column_IR")
+        card_title = self.card.content_item.title
+
+        self.assertIn(card_title, review_column)
+        self.assertNotIn(card_title, ip_column)
```
Comment (Ngoakor12) [inline backend/frontend/tests/test_card_request_review_button.py:76]: I think the functionality is missing a test for checking if the right action is logged as mentioned [here](https://github.com/Umuzi-org/Tilde/pull/679#discussion_r1477965261) on the other PR

**Review by Ngoakor12 — CHANGES_REQUESTED** [review]
This implementation seems much cleaner to me :smile: :ok_hand: 


> As suggested in previous reviews, I'm proposing we use the check_[verb]_[optional_qualifier] for naming our permission checking functions.

I think the naming convention makes sense.

> I've used it on the "request_review" action only until we all agree on a naming style for them.

We just need to remember to create a separate task for changing the rest once this PR gets merged.

> Sheena also mentioned how we do not make it known that some of these expect a "card_id" to be present in the kwargs, but I haven't been able to come up with a clean solution so far; feel free to suggest a better way to handle this.

I can't think of a better way for making this apparent. I think the "on_card" sort of makes it known that we need a way to identify the card, same thing for team related checks.

File: backend/frontend/tests/test_card_request_review_button.py (hunk 16 above)
Comment (kingraphaii) [reply backend/frontend/tests/test_card_request_review_button.py:76]: Missed that one, thanks!

**Review by Ngoakor12 — APPROVED** [review]

File: backend/frontend/templates/frontend/user/board/js_exec_action_card_moved.html [hunk 17]
```diff
@@ -0,0 +1,5 @@
...
+  moveCardToCorrectColumn("{{card.id}}", "{{card.status}}");
+</script>
```
Comment (sheenarbw) [inline backend/frontend/templates/frontend/user/board/js_exec_action_card_moved.html:5]: There is a file called view_partial_action_card_moved. It's content is:

```
{% include "frontend/user/board/partial_card.html" %}

<script>
    moveCardToCorrectColumn("{{card.id}}","{{card.status}}")
</script>
```

It is exactly the same as this file.

Can you please make it DRY

File: backend/curriculum_tracking/models.py [hunk 18]
```diff
@@ -1667,6 +1667,35 @@ def request_user_is_assignee(self):
 
         return self.assignees.first() == user
 
+    def request_user_can_request_review(self, user=None):
+        """
+        Check if current user can request review for this card
+        """
+        if self.status not in [AgileCard.IN_PROGRESS, AgileCard.REVIEW_FEEDBACK]:
```
Comment (sheenarbw) [inline backend/curriculum_tracking/models.py:1677]: would it not make sense to return false if this is not a project card? Because nobody can request a review on a topic?

like: 
```
if card.content_type_nice !== project:
    return False
```

This would simplify code elsewhere and make this function more explicit and cohesive.

File: backend/frontend/templates/frontend/user/board/partial_card.html [hunk 19]
```diff
@@ -166,5 +166,17 @@
             >Start
         </button>
     {% endif %}
+    
+    {% if card.content_type_nice == "project" %}
+        {% if card.request_user_can_request_review %}
```
Comment (sheenarbw) [inline backend/frontend/templates/frontend/user/board/partial_card.html]: Shouldn't `{% if card.request_user_can_request_review %}` be enough here? 
Ideally all the logic around figuring out if a user can request a review should be inside the `card.request_user_can_request_review` function.

Right now, most of the logic is inside the function and the rest is inside the template. It's a strange way to split things up

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Very good. This looks solid.

Just a few little issues of cleanliness.  You can send this straight back to me once you have fixed that stuff and I'll merge it.

File: backend/frontend/templates/frontend/user/board/js_exec_action_card_moved.html (hunk 17 above)
Comment (kingraphaii) [reply backend/frontend/templates/frontend/user/board/js_exec_action_card_moved.html:5]: I have removed it, had been renamed to `js_exec_[]` so we'll keep the that one

**Review by sheenarbw — APPROVED** [review]
